src/main/java/innohack/gem/service/MatchService.java

```diff
@@ -1,124 +1,177 @@
 package innohack.gem.service;
 
-import com.google.common.collect.Lists;
 import innohack.gem.dao.IGEMFileDao;
 import innohack.gem.dao.IGroupDao;
+import innohack.gem.dao.IMatchFileDao;
 import innohack.gem.entity.GEMFile;
+import innohack.gem.entity.match.MatchFileGroup;
+import innohack.gem.entity.match.MatchFileRule;
 import innohack.gem.entity.rule.Group;
 import innohack.gem.entity.rule.rules.Rule;
-import java.util.ArrayList;
-import java.util.Collection;
-import java.util.HashMap;
-import java.util.List;
-import java.util.concurrent.ConcurrentHashMap;
+import java.util.*;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.stereotype.Service;
 
 @Service
 public class MatchService {
-  private static ConcurrentHashMap<String, Collection<Group>> matchedFileGroupTable =
```

Comment:
we lost the in-memory implementation of MatchService using matchFileDao. Consider adding back the in-memory version as a sepearate IMatchFileDao. can be a seperate issue

Reply:
I haven't removed the in-memory HashMap, its at line 21-22. MatchService will load the matchFileRule and matchFileGroup from database at the start and do a back up to database when theres any change within a 2min interval.

Reply:
my bad. see wrongly